system/src/system_publish_vitals.h

```diff
@@ -0,0 +1,123 @@
+#ifndef VITALS_PUBLISHER_H
+#define VITALS_PUBLISHER_H
+
+#include <cstdarg>
+#include <cstddef>
+#include <functional>
+
+#include "system_tick_hal.h"
+
+namespace particle
+{
+
+#if (0 == PLATFORM_THREADING)
+class NullTimer
+{
+public:
+    inline static bool changePeriod(const size_t)
+    {
+        return true;
+    }
+    inline static void dispose(void)
+    {
+    }
+    inline static bool isActive(void)
+    {
+        return false;
+    }
+    inline static void reset(void)
+    {
+    }
+    inline static void start(void)
+    {
+    }
+    inline static void stop(void)
+    {
+    }
+};
+#endif // not PLATFORM_THREADING
+
+namespace system
+{
+
+/**
+ * @class VitalsPublisher system_publish_vitals.h
+ * @brief Publish vitals information
+ *
+ * Combines a protocol, function and timer, to schedule and periodically
+ * publish vitals information to the cloud. This information is then consumed
+ * by the fleet health metrics dashboard in the console.
+ *
+ * @tparam Timer An API compatible timer class with \p spark_wiring_timer.h:Timer
+ *
+ * @sa Timer
+ */
+template <class Timer>
+class VitalsPublisher
+{
+public:
+    /**
+     * @typedef publish_fn_t
+     * @brief A function requiring no parameters and returning a \p system_error_t
+     */
+    typedef std::function<int(void)> publish_fn_t;
+
+    /**
+     * @brief Constructor
+     *
+     * @param[in] publish_fn The function used to send cloud messages
+     * @param[in] timer The timer used to schedule the period
+     */
+    VitalsPublisher(publish_fn_t publish_fn, Timer* timer);
```

Comment:
That was the original implementation, but was modified to use a pointer for dependency injection to facilitate unit-testing.

Reply:
So what was the problem with testing in the original code? You can easily say that there's something wrong with the design by looking at how this class is instantiated: https://github.com/particle-iot/device-os/pull/1724/files#diff-58e1af22bce41fb40b15b3822e26ce28R69 – note how the same `spark_protocol_post_description()` function gets passed to the constructor twice, as an `std::function` and indirectly via the timer

Reply:
@sergeuz Shouldn't the implementation of `publishVitals` incorporate the `ISRTaskQueue`?

Reply:
I don't have any experience with FakeIt, but it does seem to allow modifying the behavior of existing objects: https://github.com/eranpeer/FakeIt/wiki/Quickstart#spying

I've modified one of your test cases in the following way and it seems to yield the same results:
```
particle::system::VitalsPublisher<particle::mock_type::Timer> vp;

fakeit::Mock<particle::mock_type::Timer> mock_timer(*vp.timer());

// Configure Mock
fakeit::When(Method(mock_timer, changePeriod)).AlwaysReturn(true);
fakeit::When(Method(mock_timer, dispose)).AlwaysReturn();
fakeit::When(Method(mock_timer, isActive)).AlwaysReturn(false);
fakeit::When(Method(mock_timer, reset)).AlwaysReturn();
fakeit::When(Method(mock_timer, start)).AlwaysReturn();
fakeit::When(Method(mock_timer, stop)).AlwaysReturn();

vp.disablePeriodicPublish();

fakeit::Verify(Method(mock_timer, stop));
```

The problem I have with the current code is that it still doesn't eliminate that weird decoupling and duplication of the functionality. The class still expects the external code to provide a publishing function via the Timer instance, and implements its own publishing function internally as well.

I don't see a unit test in this PR that would verify that the publisher function does indeed get called on a timer event, but besides of being essential for the functionality in this PR such a test would be a good illustration to the problem

Reply:
To address a couple of points:

First, I was able to reproduce the example you posted above, but it is only possible when the Timer is a pointer and not a first class data member - I have no idea why. I have updated the PR to allocate the Timer on the heap, and then destroy it in the destructor if it was created and not passed in.

> The class still expects the external code to provide a publishing function via the `Timer` instance

I think this is better phrased, "_The class allows..._", because the pointer is not used in production - it is only there to allow for mocking and unit testing.

> I don't see a unit test in this PR that would verify that the publisher function does indeed get called on a timer event.

That sounds more like an end-to-end or integration test, because it is really testing that the `Timer` is functioning correctly in concert with the `VitalsPublisher`. To me, it would be more interesting to know the callback was set correctly. At that point, I would expect the timer to do it's job.

That being said, it's a fair point and that relationship needs to be proven one way or another. I'll do more research to figure out what this will take to implement.